<a href="https://colab.research.google.com/github/5harad/API-201-2023/blob/main/hw/hw2/hw2-answers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# API-201 C/D PROBLEM SET #2
**Due on Thursday, September 14, at 8:00 a.m.**

**I - INSTRUCTIONS**  
To successfully complete this problem set, please follow these steps:

1. **Create a copy of this notebook in your own drive by clicking `Copy to Drive` in the menubar (this is explained below in more detail below) - *if you do not do this your work will not be saved!*** We recommend completing this problem set in Google Chrome.
2. **Insert all your answers into your copy of the document.**
    1. Please include every portion of your submission in this document unless a separate electronic file is explicitly requested.
    3. All numerical calculations should be done in the notebook itself, using R code. If you have to do calculations by hand, include a picture of your handwritten work.
    2. To add an image of handwritten work, screenshots, or anything else, add a new text cell and click the small image icon in the menubar above the cell. After inserting an image, press the `esc` key to show the image. If you have trouble with this method, upload your image to [Freeimage.host](https://freeimage.host), copy a Markdown link, and paste it in a text cell.
3. **Once your document is complete, please save and submit the notebook on Canvas as a PDF.**
    1. Click `View > Expand sections` in the menubar to make sure all work is visible (if it is greyed out then it is already visible).
    1. Click `Runtime > Run all` in the menubar to make sure all of your code is executed.
    1. Click `File > Print` in the menubar and choose the `Save as PDF` option in Chrome to export your notebook as a PDF.
    1. Check to make sure answers do not get cut off - when you go to print the PDF we recommend clicking on `More settings` and setting `Margins` to `None`.
    1. Submit your file on Canvas.


**II - IDENTIFICATION**
1. **Your Full Name:** `     `  

2. **Group Members (classmates with whom you worked on this problem set):**  
    1. `     `
    2. `     `
    3. `     `
    4. `     `
    
3. **Compliance with HKS Academic Code**  
We abide by the Harvard Kennedy School Academic code for all aspects of the course. In terms of problem sets, unless explicitly written otherwise, the norms are the following: You are free (and encouraged) to discuss problem sets with your classmates. However, you must hand in your own unique written work and code in all cases. Any copy/paste of another’s work is plagiarism. In other words, you may work with your classmate(s), sitting side-by-side (physically or remotely!) and going through the problem set question by question, but you must each type your own answers and your own code. For more details, please see syllabus.

    **I certify that my work in this problem set complies with the HKS Academic Code**
    - [ ] Yes
    - [ ] No

---

#### Load `R` libraries and data
The cell below will set up your notebook for the assignment - make sure you run it before beginning!

In addition to loading the tidyverse, this code cell will import some data - an extract from the World Bank's World Development Indicators (WDI) dataset.

In [ ]:
library(tidyverse)

# load the WDI data
wdi <- read_csv("https://5harad.com/datasets/API201/country_gdp_pop.csv")


Rows: 217 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): Country
dbl (4): GDP_1994, GDP_2019, POP_1994, POP_2019

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


# LEARNING ABOUT THE WORLD ECONOMY

The purpose of this question is two-fold: (a) to help you learn about the world economy using a widely used dataset; and (b) to ensure that you are comfortable with several R functions and operations that are commonly used to analyze data.

We will be using an extract from World Bank’s World Development Indicators (WDI), data frame `wdi`. Note that the WDI dataset reports GDP values in 2010 U.S. dollars, so that so you can directly compare values in different years. Also note that GDP and population data are not available in some years.  For this problem set, use only those observations for which data are available.

### 0. Preview the Data

Run the cell below to view the first 10 rows of the `wdi` data.

In [ ]:
head(wdi, 10)

Country             GDP_1994     GDP_2019     POP_1994 POP_2019
1  Afghanistan                   NA  21739593382 17075727 38041754
2  Albania               4791372056  14868516511  3207536  2854191
3  Algeria              89791622831 202805017529 28213774 43053054
4  American Samoa                NA           NA    52095    55312
5  Andorra               2024227689   3539838261    62677    77142
6  Angola               23311735184  98769677172 13503747 31825295
7  Antigua and Barbuda    804286308   1525046720    67204    97118
8  Argentina           274830720367 437215091207 34402672 44938712
9  Armenia               3137842747  13996188673  3283667  2957731
10 Aruba                 2091786180           NA    76700   106314

### 1. Summarize the data:

Note: for those new to R and coding, it may be helpful to revisit the R summer training available on Canvas.

**a. Use the `summary` command to get a quick idea of the distribution and range of the variables in `wdi`.**

In [ ]:
# Your answer here!

#START
summary(wdi)
#END

   Country             GDP_1994            GDP_2019            POP_1994        
 Length:217         Min.   :2.622e+07   Min.   :4.592e+07   Min.   :9.252e+03  
 Class :character   1st Qu.:3.286e+09   1st Qu.:1.174e+10   1st Qu.:5.195e+05  
 Mode  :character   Median :1.289e+10   Median :3.978e+10   Median :4.666e+06  
                    Mean   :2.259e+11   Mean   :4.659e+11   Mean   :2.605e+07  
                    3rd Qu.:1.016e+11   3rd Qu.:2.481e+11   3rd Qu.:1.462e+07  
                    Max.   :9.950e+12   Max.   :1.827e+13   Max.   :1.192e+09  
                    NA's   :38          NA's   :39          NA's   :2          
    POP_2019        
 Min.   :1.165e+04  
 1st Qu.:7.778e+05  
 Median :6.661e+06  
 Mean   :3.540e+07  
 3rd Qu.:2.544e+07  
 Max.   :1.398e+09  
 NA's   :1          

**b. From this summary output, what is the mean GDP in 2019? How many countries have missing data for 1994 population?**

Edit this cell to provide your answer!

# START
mean GDP in 2019 is $466 billion

2 countries have missing population data in 1994
#END

**c. Calculate and report total world GDP and world population in 2019 using the `sum()` function. Add the argument `na.rm = TRUE` to remove missing values.**

_Hints:_

* Recall that `sum()` takes a numerical vector as its first argument. You can extract a column from a data frame using `$` (e.g., `dataset$var1` will reference the `var1` column of your data frame `dataset`).
* You can add an argument by putting a comma and the argument in parentheses (e.g., `fun(input, arg)` includes argument `arg` to function `fun(input)`).

In [ ]:
# Your answer here!

#START
sum(wdi$GDP_2019, na.rm = TRUE)
sum(wdi$POP_2019, na.rm = TRUE)
#END

[1] 8.293351e+13

[1] 7646445824

As part of the [tidyverse](https://tidyverse.org/) package, `dplyr`'s pipe symbol `%>%` is used to perform sequential operations on a dataset. For example, below we display the GDP of Brazil in 1994 and 2019.


In [ ]:
# Run this cell

wdi %>%
  filter(Country=='Brazil') %>%
  select(GDP_1994, GDP_2019)

GDP_1994     GDP_2019    
1 1.325135e+12 2.347238e+12

In [ ]:
# You can also save output as an object for future use

brazil_gdp <- wdi %>%
  filter(Country=='Brazil') %>%
  select(GDP_1994, GDP_2019)

brazil_gdp

GDP_1994     GDP_2019    
1 1.325135e+12 2.347238e+12

**d. Now identify the five countries with the largest populations in 2019 and their respective populations in 2019 using `head()` and the tidyverse function `arrange()`.**

_Hints:_
* Use pipes `%>%` to perform multiple sequential operations on the `wdi` dataset.
* `arrange(data, desc(x))` sorts data in descending order of `x` rather than ascending order.
* `head(data, n)` returns just the first `n` rows of data.
* Make sure to use `head()` and `arrange()` in the correct order, otherwise you'll order only the first 5 countries in the dataset.


In [ ]:
# Your answer here!

# START
wdi %>%
    arrange(desc(POP_2019)) %>%
    select(Country, POP_2019) %>%
    head(5)
# END

Country       POP_2019  
1 China         1397715000
2 India         1366417754
3 United States  328239523
4 Indonesia      270625568
5 Pakistan       216565318

**e. Identify the five countries with the largest GDPs in 2019 and their respective GDPs in 2019.**

In [ ]:
# Your answer here!

# START
wdi %>%
    arrange(desc(GDP_2019)) %>%
    select(Country, GDP_2019) %>%
    head(5)
# END

Country       GDP_2019    
1 United States 1.827317e+13
2 China         1.153716e+13
3 Japan         6.210698e+12
4 Germany       3.959451e+12
5 France        2.971919e+12

### 2. Manipulate the data:

Calculate GDP per capita for each country in the database in both 1994 and 2019 by creating a new dataset called `wdi_percap` that includes two new columns called `gdp_percap_1994` and `gdp_percap_2019`, respectively.

_HINT:_ Use `mutate()` to create the new columns.

**a. What is the simple average (`mean`) of the country-level GDP per capita values in 1994 and 2019?**

_HINT:_ You can remove missing values with `mean()` or `median()` just like with `sum()` using `na.rm=TRUE`

In [ ]:
# Your answer here!

# START
wdi_percap <- wdi %>%
    mutate(gdp_percap_1994 = GDP_1994 / POP_1994,
           gdp_percap_2019 = GDP_2019 / POP_2019)

mean(wdi_percap$gdp_percap_1994, na.rm = TRUE)
mean(wdi_percap$gdp_percap_2019, na.rm = TRUE)
# END

[1] 10734.6

[1] 14867.75

**b. What is the median country-level GDP per capita in 1994 and 2019?**

In [ ]:
# Your answer here!

# START
median(wdi_percap$gdp_percap_1994, na.rm = TRUE)
median(wdi_percap$gdp_percap_2019, na.rm = TRUE)
# END

[1] 3089.98

[1] 6058.013

**c. What drives the differences between (a) and (b)?**

Edit this cell to provide your answer!

# START

While half of countries have GDP per capita below $6,058, the presence of a number of countries with much higher GDP per capita substantially increases the mean while not impacting the median.

#END



**d. Using data on 2019, what is the total population of all countries with GDP per capita below the world mean calculated in part (a)?  What fraction of the total world population does this number represent?**

There are many possible ways to solve this question, so consider the following outline as one possible way to solve the problem, but not the only solution!
1. Start by saving the mean per capita GDP to a variable
1. Create a new dataset that only contains country information for those countries below the mean.
1. Compute the total population for this subset of countries, as well as for the full set of countries, saving the results in new variables.
1. Compute the requested fraction.

In [ ]:
# Your answer here!

# START
mean_gdp <- mean(wdi_percap$gdp_percap_2019, na.rm = TRUE)

below_mean <- wdi_percap %>%
  filter(gdp_percap_2019 < mean_gdp)

pop_below_mean <- sum(below_mean$POP_2019, na.rm = TRUE)
pop_total <- sum(wdi$POP_2019, na.rm = TRUE)
share_below <- pop_below_mean / pop_total

pop_below_mean
share_below
# END

[1] 6145234509

[1] 0.803672

**e. What is your answer to (d) telling you in terms of how appropriate mean world GDP per capita is in characterizing the economic well-being of the average person in the world? [2-3 sentences]**

Edit this cell to provide your answer!

# START

From (d), over 80% of the world's population lives in countries with GDP per capita below mean GDP per capita. Therefore, the global mean GDP likely over-estimates the average person's well-being.

# END

Missing data is, unfortunately, a fact of life, and we face it here.  While we cannot necessarily fix the problem of missing data, it is important to consider its effects.

**f. Use the `is.na` function to determine the number of countries for which the WDI database does not have a 2019 GDP figure.**

_HINT:_ The function `is.na()` returns `TRUE` for missing values and `FALSE` otherwise. You can calculate the total number of missing values in a vector `x` using `sum(is.na(x))`.

In [ ]:
# Your answer here!

# START
sum(is.na(wdi$GDP_2019))
# END

[1] 39

### 3. Merging data

We now call in two more datasets from `csv` files:   `co2`, which includes information on country CO2 emissions (metric tons), and `continent` which includes the continent of each country.

In [ ]:
# Run this cell
co2 <- read_csv("https://5harad.com/datasets/API201/country_co2.csv")
continent <- read_csv("https://5harad.com/datasets/API201/country_continent.csv")

Rows: 209 Columns: 2
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): Country
dbl (1): CO2_2016

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 191 Columns: 2
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): Country, Continent

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


Preview the first few observations of each dataset.

In [ ]:
# Run this cell

# START
head(co2, 10)
head(continent, 10)
# END

Country             CO2_2016 
1  Afghanistan           9900004
2  Albania               5208319
3  Algeria             156220560
4  Angola               30566933
5  Anguilla                30262
6  Antigua and Barbuda    438763
7  Argentina           200708270
8  Armenia               4597845
9  Aruba                  286871
10 Australia           414988700

Country             Continent    
1  Afghanistan         Asia         
2  Albania             Europe       
3  Algeria             Africa       
4  Angola              Africa       
5  Antigua and Barbuda North America
6  Argentina           South America
7  Armenia             Asia         
8  Australia           Oceania      
9  Austria             Europe       
10 Azerbaijan          Asia

A `left_join` command can merge the new datasets with `wdi`. The join requires a common variable, which is `Country` in this case.

Let's first do this for `co2`.

In [ ]:
# Run this cell

wdi_with_co2 <- left_join(wdi, co2, by = "Country")
head(wdi_with_co2, 10)


Country             GDP_1994     GDP_2019     POP_1994 POP_2019 CO2_2016 
1  Afghanistan                   NA  21739593382 17075727 38041754   9900004
2  Albania               4791372056  14868516511  3207536  2854191   5208319
3  Algeria              89791622831 202805017529 28213774 43053054 156220560
4  American Samoa                NA           NA    52095    55312        NA
5  Andorra               2024227689   3539838261    62677    77142        NA
6  Angola               23311735184  98769677172 13503747 31825295  30566933
7  Antigua and Barbuda    804286308   1525046720    67204    97118    438763
8  Argentina           274830720367 437215091207 34402672 44938712 200708270
9  Armenia               3137842747  13996188673  3283667  2957731   4597845
10 Aruba                 2091786180           NA    76700   106314    286871

**a. Now merge `continent` into your dataset.**

Note: some countries won't match because different datasets can have different ways of spelling/naming a country. This is OK for now, but in future research, ensuring consistency is an important part of the data cleaning process!

In [ ]:
# Your answer here!

# START
wdi_augmented <- left_join(wdi_with_co2, continent, by = "Country")
head(wdi_augmented, 10)
# END

Country             GDP_1994     GDP_2019     POP_1994 POP_2019 CO2_2016 
1  Afghanistan                   NA  21739593382 17075727 38041754   9900004
2  Albania               4791372056  14868516511  3207536  2854191   5208319
3  Algeria              89791622831 202805017529 28213774 43053054 156220560
4  American Samoa                NA           NA    52095    55312        NA
5  Andorra               2024227689   3539838261    62677    77142        NA
6  Angola               23311735184  98769677172 13503747 31825295  30566933
7  Antigua and Barbuda    804286308   1525046720    67204    97118    438763
8  Argentina           274830720367 437215091207 34402672 44938712 200708270
9  Armenia               3137842747  13996188673  3283667  2957731   4597845
10 Aruba                 2091786180           NA    76700   106314    286871
   Continent    
1  Asia         
2  Europe       
3  Africa       
4  NA           
5  NA           
6  Africa       
7  North America
8  South America
9  Asia         
10 NA

**b. Add a column to this joined dataset indicating CO2 emissions per capita — called `co2_emissions_pc` — using 2019 population. What is the mean value?**

In [ ]:
# Your answer here!

# START

wdi_augmented <- wdi_augmented %>%
  mutate(co2_emissions_pc = CO2_2016/POP_2019)

mean(wdi_augmented$co2_emissions_pc, na.rm = TRUE)
# END

[1] 4.545887

**c. Which five countries have the highest CO2 emissions per capita? Which five have the lowest?**

In [ ]:
# Your answer here!

# START

# Highest
wdi_augmented %>%
    arrange(desc(co2_emissions_pc)) %>%
    select(Country, co2_emissions_pc) %>%
    head(5)

# Lowest
wdi_augmented %>%
    arrange(co2_emissions_pc) %>%
    select(Country, co2_emissions_pc) %>%
    head(5)

# END

Country              co2_emissions_pc
1 Qatar                34.95330        
2 Montenegro           26.11810        
3 Trinidad and Tobago  25.07164        
4 Kuwait               24.12413        
5 United Arab Emirates 22.39272

Country   co2_emissions_pc
1 Greenland 0.02721209      
2 Mali      0.08171831      
3 Somalia   0.08213753      
4 Niger     0.08959292      
5 Ethiopia  0.09313859

**d. Calculate total CO2 emissions for each continent. Order the continents from highest to lowest.**

HINT: this can be done with the `group_by()` operation.

In [ ]:
# Your answer here!

#START
wdi_augmented %>%
    group_by(continent) %>%
    summarize(CO2_2016 = sum(CO2_2016, na.rm = TRUE)) %>%
    arrange(desc(CO2_2016))
# END

continent     CO2_2016   
1 Asia          17702259465
2 North America  6229033322
3 Europe         4204428521
4 South America   992738341
5 Africa          974111086
6 Oceania         460050401
7 NA               46560108

**e. Compute the correlation between CO2 emissions per capita and GDP per capita in 2019**

*HINT*: to deal with missing values, you can add `use="complete.obs"` as an argument in `cor()`.



In [ ]:
# Your answer here!

#START
wdi_augmented %>%
    mutate(gdp_percap_2019 = GDP_2019 / POP_2019) %>%
    summarize(cor = cor(co2_emissions_pc, gdp_percap_2019, use="complete.obs"))
# END

cor      
1 0.5755508

**f. What do you think is driving this relationship? Why? Qualitative answers are fine.**

Edit this cell to provide your answer!

# START
0.57 is a relatively high correlation in the real world. This is likely driven by the fact that richer countries consume more energy and energy-intensive products, and thus have higher emissions per capita.
# END